### PR TITLE
[767] Fix dead trades when they should exists (potential hotfix candidate)

### LINKED ISSUE #767: `null` trades where they shouldn't be
**Description**
"Original trade", created using the Uni useTrade hook is used to cheaply populate our trade extended object with the uni trade methods e.g executionPrice etc. but it is returning `null` on certain token pairs when called and thus returns null on our trade object despite valid fee/price calls returning data
image

**How to Reproduce**
1. Mainnet
2. import CVX token 0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b
3. input amt in any field
4. open console to networks and see the calls being made normally and returning data
5. inifinite loading state (error)

**Expected behavior**
Should return a trade as there is indeed liquidity and the backend returns us data for those params using CVX

**Additional context**
We will need to rewrite our own methods. For now I suggest just copying them from Uni's Trade sdk until we can do a proper rewrite

slack: https://gnosisinc.slack.com/archives/CPZA1AGKY/p1623671889000800

### PR DESCRIPTION
**Pull request by W3stside**
Closes #767 
Conversation for more details: https://gnosisinc.slack.com/archives/CPZA1AGKY/p1623671889000800

Problem: I had coded our Trade extension code to use the original Uni `useTradeExactIn` and `useTradeExactOut` hooks, which create a uni `Trade` entity in order to get the remaining methods as part of the `Trade` type, to avoid having to 
a) create our own custom sdk
and
b) change all the types afterwards (assumption is it our own `TradeGp` entity would differ from the uni `Trade` type)

but now we have to make this change as it is breaking some trades, in this case CVX <> DAI on mainnet. Trade exists from our backend but NOT from Uni, inside the `useTradeExact(in/out)` hook, returning `null` for the entire trade extension calculation.

That's obviously bad. 

This PR creates our own simplified sdk using the SAME LOGIC as before, but into our own class which i then fixed the broken types everywhere via mods and the like.

Mods:
1. AdvancedSwapDetailsDropdown
2. utils/prices

### Testing
- testing this is a bit tough but I did it myself by logging what the output of using the uni hook `useTradeExactIn` and `useTradeExactOut` return values vs what the backend was returning. Uni hooks were always `null` on certain token pairs while the backend wasn't and thus no trades.
- this no longer happens with this new trade addition

### CODE REVIEW

File: src/custom/state/swap/TradeGp.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
+import { CanonicalMarketParams, getCanonicalMarket } from '@src/custom/utils/misc'
```
Comment (W3stside) [inline src/custom/state/swap/TradeGp.ts:1]: the majority of the logc (unchanged logically) that WAS in `trade/extensions.ts` is now here as a class with the necessary methods coded in. This solves the need for the previous uni trade to be mixed in to meet type requirements.

File: src/custom/state/swap/TradeGp.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  }
+  /**
+   * Get the maximum amount in that can be spent via this trade for the given slippage tolerance
+   * @param slippageTolerance tolerance of unfavorable slippage from the execution price of this trade
+   */
+  public maximumAmountIn(slippageTolerance: Percent): CurrencyAmount {
+    return _maximumAmountIn(slippageTolerance, this)
+  }
+}
+export {}
```
Comment (alfetopito) [inline src/custom/state/swap/TradeGp.ts:151]: Hum?

**Review by alfetopito — APPROVED** [review]
Tested out, working great.

File: src/custom/state/swap/TradeGp.ts (hunk 2 above)
Comment (W3stside) [reply src/custom/state/swap/TradeGp.ts:151]: oops

Comment (W3stside) [reply src/custom/state/swap/TradeGp.ts:151]:
```suggestion
```

**Comment by W3stside** [comment]
@alfetopito wondering if we should wait until wednesday when @anxolin is back to merge this as its pretty pressing?

**Comment by alfetopito** [comment]
> @alfetopito wondering if we should wait until wednesday when @anxolin is back to merge this as its pretty pressing?

No, we should go ahead with this one as it's a pretty big fix. You can address any comments Anxo might have in a follow up PR.

**Pull request merged**